src/content/blog/2024_qcoder.md

```diff
@@ -0,0 +1,163 @@
+
+First, apply the Hadamard gate to the first quantum bit.
+$$
+\begin{equation}
+\ket{000 \cdots 0} \xrightarrow{H(0)}  \frac{1}{\sqrt{2}} \lparen \ket{000 \cdots 0} + \ket{100 \cdots 0} \rparen
+\end{equation}
+$$
+Next, we transform the state $\ket{100 \cdots 0}$ to $\ket{110 \cdots 0}$.
+To achieve this, apply a controlled-X gate (CNOT gate) with the first quantum bit as the control bit and the second quantum bit as the target bit:
+$$
+\begin{equation}
+\frac{1}{\sqrt{2}} \lparen \ket{000 \cdots 0} + \ket{100 \cdots 0} \rparen \xrightarrow{CX(0,1)}  \frac{1}{\sqrt{2}} \lparen \ket{000 \cdots 0} + \ket{110 \cdots 0} \rparen
+\end{equation}
+$$
+By continuing to apply the controlled-X gate until the $n$-th qubit is the target bit, we can prepare the GHZ state.
+$$
+\begin{equation}
+\frac{1}{\sqrt{2}} \lparen \ket{000 \cdots 0} + \ket{110 \cdots 0} \rparen \xrightarrow{CX(0,2)} \, \cdots \xrightarrow{CX(0,n-1)} \frac{1}{\sqrt{2}} (\ket{0...0} + \ket{1...1})
+\end{equation}
+$$
+Summarizing these operations, we obtain the following circuit when $n = 4$:
+
+<p align="center">
+    <img src='../../images/2024-qcoder/QPC002_A3_circuit.png' width=500/>
+</p>
+
+Below is a sample program:
+```python
+from qiskit import QuantumCircuit
+ 
+ 
+def solve(n: int) -> QuantumCircuit:
+    qc = QuantumCircuit(n)
+ 
+    qc.h(0)
+ 
+    for i in range(1, n):
+        qc.cx(0, i)
+ 
+    return qc
+```
+
+</details>
+
+## Solution 2
+<details>
+<summary>Open</summary>
+
+In the solution 1, the depth of the quantum circuit is $n$.
+What kind of quantum circuit should be designed to further reduce the depth of the quantum circuit?
+
+Let's examine the following example with 8 qubits.
+
+<p align="center">
+    <img src='../../images/2024-qcoder/QPC002_A5_circuit.png' width=600/>
+</p>
+
+In this case, the operations within each block, divided by the boundaries on the quantum circuit, act on different qubits, allowing them to be executed simultaneously. Therefore, the depth of each block is 1.
+Thus, the depth of the quantum circuit above is 4.
+
+By generalizing this circuit, the depth of the quantum circuit becomes $\lceil \log_{2}{n} \rceil + 1$, which satisfies the given constraint.
+
+Below is a sample program:
+```python
+import math
+
+from qiskit import QuantumCircuit
+
+
+def solve(n: int) -> QuantumCircuit:
+    qc = QuantumCircuit(n)
+
+    qc.h(0)
+
+    for i in range(int(math.log2(n)) + 1):
+        for j in range(2**i):
+            if 2**i + j == n:
+                break
+            qc.cx(j, 2**i + j)
+
+    return qc
+```
+
+</details>
+
+## Note
+Almost the same problem was featured in the <a href="https://www.qcoder.jp/contests/QPC002/" target="_blank">QCoder Programming Contest 002</a>. In Problem A3, both solutions 1 and 2 are accepted. However, in Problem A5, only Solution 2 is accepted, as it meets the stricter depth constraints. If you're interested, feel free to submit your solution and see how you do!
```

Comment:
For all I know, the circuit you've constructed above might be the theoretical limit when creating the GHZ state from $\{H, S, T, CX\}$. BTW, are you restricting code submissions to use a specific gate set?
```suggestion
Almost the same problem was featured in the <a href="https://www.qcoder.jp/contests/QPC002/" target="_blank">QCoder Programming Contest 002</a>. In Problem A3, both solutions 1 and 2 are accepted. However, in Problem A5, only Solution 2 is accepted, as it meets the stricter depth constraints. If you're interested, feel free to submit your solution and see how you do! Can you find a shorter circuit?
```

Reply:
Currently, users can use Qiskit's [standard gates](https://docs.quantum.ibm.com/api/qiskit/circuit_library#standard-gates) in their code **unless the problem specifies otherwise:** https://www.qcoder.jp/en/qa#gate I think stating this aspect in the sample problem may be too much and could potentially confuse the reader. What are your thoughts?